arch/riscv/mm/mem_encrypt.c

```diff
@@ -25,25 +25,40 @@ bool force_dma_unencrypted(struct device *dev)
 
 int set_memory_encrypted(unsigned long addr, int numpages)
 {
+	int i, rc;
+
 	if (!cc_platform_has(CC_ATTR_MEM_ENCRYPT))
 		return 0;
 
 	if (!PAGE_ALIGNED(addr))
 		return -EINVAL;
 
-	return sbi_covg_unshare_memory(__pa(addr), numpages * PAGE_SIZE);
+	rc = sbi_covg_unshare_memory(__pa(addr), numpages * PAGE_SIZE);
+	if (rc)
+		for (i = 0; i < numpages && rc == 0; i++)
+			rc = sbi_covg_unshare_memory(__pa(addr + i * PAGE_SIZE), PAGE_SIZE);
+
+	return rc;
```

Comment:
CoVE implementations with support for Smmtt hardware can move individual pages between confidential and non-confidential memory. With this change, we add support for CoVE implementations that do not support Smmtt hardware. In such implementations, TSM might return error when a TVM tries to share/unshare more than a single page at the time. The fallback is to share/unshare page by page.